### ISSUE TITLE
Question: the logic behind the listen() loop in JustSayingBus.cs:86

### ISSUE THREAD

**Issue by dariusdamalakas** [justeat/JustSaying#77]
Hi,

This is more of a question not an issue. It's note exactly clear from docs whether we can have handlers for more than one message. I assume it is, so this is what it looks like:
```cs
CreateMeABus.InRegion(RegionEndpoint.EUWest1.SystemName)
             .WithSqsTopicSubscriber()
             .IntoQueue(queue)
             .WithMessageHandler<SubscriptionAdded>(new SubscriptionHandler())
             .WithMessageHandler<SubscriptionRemoved>(new SubscriptionHandler())
             .StartListening();
```

Now, looking at [JustSaying.Bus.cs:86](https://github.com/justeat/JustSaying/blob/master/JustSaying/JustSayingBus.cs), the Start() method would have 2 subcsribers, because there are 2 subscribers by topic. However, both topic subscribers are listening for the same queue, so is it the case that both listeners will poll the same queue at same time?

This all makes sense for multiple regions - needs to listen for each region separately as been noted in [issue 59](https://github.com/justeat/JustSaying/issues/59). But this case is just for single region single queue.

I might have completely misunderstood how configuration is supposed to be used for multiple handlers.

Thanks

**Comment by dariusdamalakas** [comment]
I've looked into this further, and I can see that listener per topic paradigm is core to JustSaying. 

There is a [unit test](https://github.com/justeat/JustSaying/blob/master/JustSaying.UnitTests/JustSayingBus/WhenRegisteringMessageHandlers.cs) which explicitly test that multiple listeners will be started for each topic, regardless of whether topic listens into the same queue or not. Line 60 and 61 both start listeners.

**Comment by dariusdamalakas** [comment]
More info. The reason I am chasing this is because having 2 listeners listening to same queue actually impacts message delivery in a negative way. I've written a sample consumer and producer app. Out of 200 messages published, 10 to 20 messages are not handled correctly and are sent to DLQ. This is strange, even with 2 listeners polling SQS queue that should still work correctly, there probably must be some threading issue.  If I change sample to just have 1 handler (and 1 listener), then things go back to normal and 200 messages are processed significantly more quicker, on my specific case, at least twice or so.    Can post the sample if it's needed.

**Comment by payman81** [comment]
Multiple listener per queue has been supported from the start but admittedly so far I've stuck to using single listener per queue for reasons other than performance e.g. uncomplicated retry policy when one handler fails.

If you want to have different queues for your handlers you can use the following syntax:
```cs
CreateMeABus.InRegion(RegionEndpoint.EUWest1.SystemName)
             .WithSqsTopicSubscriber()
             .IntoQueue(subscriptionAddedQueue)
             .WithMessageHandler<SubscriptionAdded>(new SubscriptionHandler())

             .WithSqsTopicSubscriber()
             .IntoQueue(subscriptionRemovedQueue)
             .WithMessageHandler<SubscriptionRemoved>(new SubscriptionHandler())
             .StartListening();
```

**Comment by dariusdamalakas** [comment]
Ok, so just to understand - there is no requirement for having multiple listeners for the same queue?

I've made a quick change yesterday to register INotificationListeners per region per queue, instead of per-topic, that seems to work fine. Would that be accepted as a PR? 

Just to be clear - your example would still work, as there would be 2 queues in 1 region.

**Comment by payman81** [comment]
No the library doesn't force you to have multiple listeners for the same queue.
You can extend the code above and have as many queues as you want.

**Comment by payman81** [comment]
Is this what you wanted or there is still a scenario that JustSaying does not seem to accomodate?

**Comment by dariusdamalakas** [comment]
>>No the library doesn't force you to have multiple listeners for the same queue.
I don't find this to be the case though.  Library does start multiple listeners for the same queue for each handler that we register (ignoring the regions here, just 1 region). Sorry if I am not expressing myself clear here.

**Comment by payman81** [comment]
In the code snippet that I posted earlier, you have 2 queues. One for SubscriptionAdded event and one for SubscriptionRemoved event. Each queue has only a single listener i.e. subscriptionHandler.

Is that not the case?

**Comment by dariusdamalakas** [comment]
Yes, that's the case, but this is not my point. The point is - why there are multiple listeners polling the same queue?

**Comment by payman81** [comment]
Sorry I'm still not clear where the multiple listeners polling the same queue come from in the code snippet that I shared. I'll ask other colleagues to weigh in as well. CC @brainmurphy / @antonjefcoate

**Comment by payman81** [comment]
Also, it would help if you can share your spike or post a minimal complete sample. Cheers.

**Comment by dariusdamalakas** [comment]
>Sorry I'm still not clear where the multiple listeners polling the same queue come from in the code snippet that I shared

That's because it's happening in this snippet, not yours.    One handler per queue is fine - as that will fire up just one listener. If you have 2 handlers, as in the snippet below, 2 listeners will be fired. They will both listen to queue1.   

```
CreateMeABus.InRegion(RegionEndpoint.EUWest1.SystemName)
             .WithSqsTopicSubscriber()
             .IntoQueue(queue1)
             .WithMessageHandler<SubscriptionAdded>(new SubscriptionAddedHandler())
             .WithMessageHandler<SubscriptionRemoved>(new SubscriptionRemovedHandler())
             .StartListening();
```

**Comment by dariusdamalakas** [comment]
I've read what i've written before, and I might not have expressed myself clear.  I'll reword it:

There is a [unit test](https://github.com/justeat/JustSaying/blob/master/JustSaying.UnitTests/JustSayingBus/WhenRegisteringMessageHandlers.cs) which explicitly test that a listener will be started for each topic. This results in 2 listeners polling same SQS queue. See line 60 and 61 that check for 2 calls to Listen().

**Comment by brainmurphy** [comment]
That unit test does indeed show that you can have two handlers listening to the same queue - that's the behaviour you see when you use the fluent syntax to say:

```
	.WithSqsTopicSubscriber()
	.IntoQueue(queue1)
	.WithMessageHandler<Message1>(handler1)
	.WithMessageHandler<Message2>(handler2)
```

As you correctly observe, that *does* result in two listeners. JustSaying allows you to do that, therefore preventing that would be a breaking change.

JustSaying, however, doesn't force you to do that. You can say:

```
	.WithSqsTopicSubscriber()
	.IntoQueue(queue1)
	.WithMessageHandler<Message1>(handler1)
	.WithSqsTopicSubscriber()
	.IntoQueue(queue2)
	.WithMessageHandler<Message2>(handler2)
```

Which, as you've seen, results in two queues, each subscribed to a different topic and each with a different listener.

Preventing the first behaviour would be a breaking change and some clients may be relying on that behaviour (though it's not a pattern I use for the reasons @payman81 alluded to).

Does that make sense?

**Comment by brainmurphy** [comment]
Is there a specific scenario, @dariusdamalakas, that you think that JustSaying is unable to support right now? Or are you interested in an explanation for the current behaviour (which I've hopefully helped with)?

**Comment by dariusdamalakas** [comment]
Thanks @brainmurphy, all fine with the multiple queues scenario. I appreciate this is a breaking change. To be fair, I did expect this kind of answer, as I can see this was a deliberate decision to have a listener per topic. don't see any usages for that, but it is what it is.

There's a [PR#77](https://github.com/justeat/JustSaying/pull/78) which shows how we've modified JustSaying to have a listener per queue instead of listener per handler. Feel free to decline this if this change is breaking and this can not be brought in.

**Comment by dariusdamalakas** [comment]
Hi,

>Preventing the first behaviour would be a breaking change and some clients may be relying on that behaviour

Actually, I don't think this is working. I've created a minimal sample, [source code here](https://etherpad.mozilla.org/ep/pad/view/ro.ums0EqRuwF7/latest).   Copy into new empty project, install latest JustSaying, and fire it up. The sample will publish 1000 messages, and will consume them. There are two methods where you can switch between having 1 listener per queue, or 2 listeners per queue. Since we are publishing just 1 type of message in both cases, that's good enough for an example.

I've chosen rather high number for messages published so that it would trigger odd behaviour. I get more or less 50 messages sent to DLQ every time I run this with StartTwoListenersPerQueue(). If i use one listener on a queue StartOneListenersPerQueue(), all messages are received fine. Please alternate between these two and let me know if you see any difference in behaviour.

**Comment by payman81** [comment]
@dariusdamalakas Indeed the sample with 2 handlers is losing messages. This bug must've been re-introduced following relatively recent updates. 
Just to clarify, this bug only impacts pub-sub model not point-to-point model.

**Comment by dariusdamalakas** [comment]
Thanks. Glad we finally we are finally on the same page. Just to add for the record. PR #78 was specifically for addressing this issue. Now, the way it was solving this problem was just by limiting to 1 listener per queue. This obviously just hides the issue, and prevents the bug from manifesting. It solves the issue, but not by actually fixing the bug. Hence i'm happy to close that PR.

On another record, I do think that having multiple listeners on the same queue might be desired in some cases, such as when 1 listener can not empty the queue as fast as possible, and there is still potential increase the throughput by having `N` listeners polling same queue. This would be very neat.

Just to brainstorm: I think one idea to allow to configure the number of listeners per queue would be to introcuce a 'WithNumberOfListeners(int)` configuration option. This would obviously default to 1 listener per queue, but would also allow to increase that on a per-queue basis.

**Comment by dariusdamalakas** [comment]
Found the root cause of this issue. The issue is caused by the fact that listeners are created per topic they are 'listening'. I've putting this in quotes, because listeners poll queues, not topics. 

So to trace further how the issue manifests itself, we can see what happens when message handlers are being added in [JustSayingBus.cs:64](https://github.com/justeat/JustSaying/blob/master/JustSaying/JustSayingBus.cs#L64). Here's the actual code:

````
public void AddMessageHandler<T>(Func<IHandler<T>> futureHandler) where T : Message
{
    var topic = typeof(T).ToTopicName();
    foreach (var subscriber in _subscribersByTopic[topic])
         subscriber.AddMessageHandler(futureHandler);
}
````

Subscriber is found by getting topic name from message type, and then handler is added to that subscriber. This means if we register 2 handlers for same queue for, as an example, messages A and B, we will have 2 subscribers, each subscriber having only 1 handler registered, i.e. first subscriber will have handler for message A, second subscriber wit handler for message B.

Now, as both of these subscribers poll the same queue, they will receive messages A and B, because polling is done on a queue level. But as we saw, each subscriber will have only one handler, either for message A or B. If we continuously publish messages A, eventually, that message will be picked up by second subscriber. As a result, messages A will not be processed by second subscriber, because subscriber only handles messages B. Message processing loop will be discontinued at this location [SqsNotificationListener.cs:172](https://github.com/justeat/JustSaying/blob/fcb9e827c52cf93560bd69ebadfb4a76cbfd4d15/JustSaying.AwsTools/SqsNotificationListener.cs#L172)

**Comment by dariusdamalakas** [comment]
@brainmurphy, @payman81 , i'd really like to agree for a solution on this, we already have a number of changes done in our [fork](https://github.com/Intelliflo/JustSaying), and we would like to contribute them back. 

I suggest to fix this in 2 steps:
1. First we need to fix this bug. You can see the core of the fix in file [JustSaying.Bus file on listener-per-queue branch](https://github.com/Intelliflo/JustSaying/blob/listener-per-queue-per-region/JustSaying/JustSayingBus.cs#L76).  
 Having multiple listeners per queue is not working in its current state. I would argue whether someone is using multiple listeners on queue, because, as we've seen, the design of this has issues which prevent subscribers from handling messages. This leads to say that PR #78 can be safely applied without affecting any existing 'JustSaying' users functionality.
2. Second step is to introduce a configuration option which would allow to configure the number of listeners per queue. 

A possible configuration could look like this.

````
CreateMeABus.InRegion(RegionEndpoint.EUWest1.SystemName)
                .WithSqsTopicSubscriber()
                .IntoQueue(queue)
                .WithMessageWorkers(5)
````

The call to `.WithMesageWorkers(5)` would define the number of listeners fired for queues. This can be raised as a separate enhancement issue on github.

**Comment by payman81** [comment]
@dariusdamalakas Your root cause analysis sounds plausible. I'd like to get this issue resolved as soon as possible by fixing the root cause. 
I think the number of handlers polling the queue should be implied when you register the handlers rather than explicitly set as in `.WithMesageWorkers(5)`. 
What I expect from the fluent syntax below is a single queue with 2 handlers since there are 2 handlers registered for a given SqsTopicSubscriber.
````
CreateMeABus.InRegion(RegionEndpoint.EUWest1.SystemName)
             .WithSqsTopicSubscriber()
             .IntoQueue(queue1)
             .WithMessageHandler<SubscriptionAdded>(new SubscriptionAddedHandler())
             .WithMessageHandler<SubscriptionRemoved>(new SubscriptionRemovedHandler())
             .StartListening();
````
My hope is that by fixing the underlying bug we can make your scenario possible without changing the fluent syntax.

**Comment by dariusdamalakas** [comment]
@payman81, can you please elaborate on what you mean exactly by 'implied'?   I do not see the reason why one would want to tie handler to a subscriber. 

From your example, I see the following things, in most part I agree with you, except point c
a) 2 handlers
b) single queue
c) single subscriber/sqs notification listener.

It's the point c) that I care most. I'd expect library to allow greatest  flexibility in configuration. Tying number of handlers to the number of subscribers rings no bell to me. Why would someone want to have 50 subscribers or so to the same queue, when clearly 1 or 2 would be able to drown the queue fast enough. 50 was just taken to illustrate the point that linking handlers with number of subscribers simply does not scale.

**Comment by payman81** [comment]
Hi @dariusdamalakas - I think there has been a misunderstanding caused by inconsistent use of language. 
For example, I've been referring to 'handlers' and 'subscribers' interchangeably which in my opinion makes sense from the point of view of users of JustSaying because conceptually, you have publishers and subscribers communicating over a message bus with most of the machinery abstracted away.

But from the point of view of developers of JustSaying, 'queue subscribers' aka listeners are responsible for polling the queue and invoking handlers. As you rightly pointed out, they are not the same thing.

That is "subscriber" == SqsNotificationListener, yes? This is responsible for polling a single queue and calling the appropriate handler for the message retrieved by that polling operation.

I do think having a single 'queue subscriber' will solve the problem but I'm not sure about allowing N subscribers. Did you actually have the case where your messages weren't being picked up because subscriber is lagging behind? Most certainly, your handlers will be the bottleneck and not the subscriber. All subscribers do is pop the message and delegate to your handlers.

**Comment by dariusdamalakas** [comment]
>That is "subscriber" == SqsNotificationListener, yes? 

Yes, correct. This is all terminology that is being used in JustSaying, i.e `INotificationSubscriber` which has just single implementation `SqsNotificationListener`.

>I do think having a single 'queue subscriber' will solve the problem 

Ok, good. Where does that bring us with regards to PR [listener-per-queue-per-region](https://github.com/Intelliflo/JustSaying/tree/listener-per-queue-per-region) ?

> but I'm not sure about allowing N subscribers.

As I said, this is a second and optional step.  As you have correctly pointed out, the most likely bottleneck will be the handlers, so having more than one subscriber can be left for out for now. I am happy with that.

**Comment by brainmurphy** [comment]
> Ok, good. Where does that bring us with regards to PR [listener-per-queue-per-region](https://github.com/Intelliflo/JustSaying/tree/listener-per-queue-per-region) ?

My original worry, as you remember, with that PR was that it represented a breaking change (multiple subscribers to the same queue). However, it looks like we're now sure that the functionality that would be lost doesn't actually work correctly: messages can get lost. Also, from talking to @payman81 offline, it sounds like we *don't* actually have anybody internally @ JUST EAT using that functionality. So: great. We're on the same page :+1:

Let's have a quick think. If we could express our intention through the fluent syntax, that'd be better; agreed? An ugliness at the moment is that that `SqsNotificationListener` is actually being created as part of `WithMessageHandler`. If we created that earlier (as part of `WithSqsTopicSubscriber` or `IntoQueue`), then we would de-couple that and the code would express what's actually happening.

Then, this would do what you expect (one queue with one subscriber with two handlers):
```csharp
CreateMeABus
    .InRegion("eu-west-1")
    .WithSqsTopicSubscriber()
    .IntoQueue("queue1")
    .WithMessageHandler(handler1)
    .WithMessageHandler(handler2);
```

As would this (two queues, each with one subscriber, each with one handler):
```csharp
CreateMeABus
    .InRegion("eu-west-1")
    .WithSqsTopicSubscriber()
    .IntoQueue("queue1")
    .WithMessageHandler(handler1)
    .WithSqsTopicSubscriber()
    .IntoQueue("queue2")
    .WithMessageHandler(handler2);
```

Interested in your opinion, @dariusdamalakas.

**Comment by dariusdamalakas** [comment]
Yes, agreed, this is how I was understanding this fluent syntax from early on. But then quickly realised that this wasn't the case, that somehow subscriber was tied in with handler, hence this lengthy thread that we are having :+1:  

So the current state of PR is that subscribers are still created as part of `WithMessageHandler` call. Granted, this obviously is not the best place, I just didn't took this PR further, as didn't see that this PR would be accepted at all. Agree that 'SqsNotificationListener` should be created as part of 'WithSqsTopicSubscriber' or 'WithSqsPointToPointSubscriber'.

**Comment by brainmurphy** [comment]
Superb. I'll spend some time today getting a PR together to reflect what we've talked about here; definitely feels like this would be an improvement to the code. Will ask you to have and @payman81  a look.

**Comment by brainmurphy** [comment]
Well, this has been a bit more problematic than anticipated (familiar story...) You'll notice that `IntoQueue` currently allows you to supply `string.Empty` as the queue name - the subsequent call to `WithMessageHandler` is generic and we can then use the type to build the queue name. This is often how JustSaying is used at JUST EAT - giving a queue-per-message type.

If I instead want to create and subscribe to the queue in `IntoQueue`, I have a problem now: what queue name do I use if string.Empty has been supplied since I don't have a message type at that point? Answer: dunno. I feel like this is definitely the way the code should behave, but it's not possible without breaking a bunch of stuff right now - unfortunate!

Which brings us back to your previous PR, @dariusdamalakas. I'll have another look at that and verify that it make sense. All of the integration tests are passing, yes?

**Comment by dariusdamalakas** [comment]
@brainmurphy , let me have another look at that PR, if that's going to be accepted, I will make sure that everything passes, and will make some more adjustments, there are some unnecessary code execution branches that need to be dropped or refactored.

I am very keen to get this PR in, as there are some other things that we want to contribute back. We are now running on our own branch of JustSaying, 17 commits ahead, and has 2 new interesting features: allows to provide custom topic/queue naming strategy via fluent configuration (fixes #79), and improved message serializer factory, which allows to handle bounce notification from AWS SES (improvement #80).

Will send a PR once i've updated it.

**Comment by dariusdamalakas** [comment]
PR ready: #89 .

Some notes on it. There is some room for improvement, especially in the area where subscribers are created. However, refactoring that here would be difficult, because as I said, we have some other changes built on this already. Specifically, issue #79 deals with queue and topic naming. We have introduced `INamingStrategy`, more on this later, but what I want to point out is that further refactorings to this PR would mean breaking changes for other pull requests we will send. Basically, what I am saying is that things will get refactored, just not exactly now, hope this makes sense.

I've ran all integration tests as well as unit tests, all passed fine.
